src/main/python/pgshovel/replication/streams/kafka.py

```diff
@@ -40,12 +54,44 @@ def consume(self, state):
             consumer.seek(delta, 1)
             assert consumer.offsets[0] == offset
 
-        for offset, message in consumer:
-            decoded = self.codec.decode(message.value)
-            state = validate_state(state, offset, decoded)
+        for offset, message in self.__primed(consumer, state):
+            state = validate_state(state, offset, message)
             # XXX: This is necessary because of a bug in protocol buffer oneof.
             state = type(state).FromString(state.SerializeToString())
-            yield state, offset, decoded
+            yield state, offset, message
+
+    def __primed(self, consumer, state, attempts=5):
+        decoded = imap(
+            lambda (offset, msg): tuple((offset, self.codec.decode(msg.value))),
+            consumer
+        )
+
+        offset, message = next(decoded)
+
+        try:
+            validate_state(state, offset, message)
+        except InvalidEventError as e:
+            # See if we can drop the first few messages in an attempt to find
+            # the type we're looking for
+            while attempts > 0:
```

Comment:
So I didn't originally go the route you mentioned because I felt like checking the stream state to see if it started with a `BeginOperation`, and then if not priming it to do so, was basically validation + steps to rectify.  And doing the validation part not with the `validation` package tools felt like bad encapsulation.  The code that primes (the stream itself) has to know that a ``BeginOperation`` is the only valid way to start a stream, which I felt wasn't its responsibility and was rather the `validation` package's responsibility . Hence why I had the `validation` package provide more info on the actual validation error, allowing the caller to fix their state as needed before continuing.

That said, points you made here and in the other linked to comment are good ones, so I'm now on the fence.  I'll think about this some more and see if I can find something I like.

P.S. another option may be to ask for the valid starting events by accessing  `replication.validation.validate_state.validators['stream_state']['transaction_state'].receivers[None].keys()`.  That can then be used with a simpler `prime` method that just does the `itertools.dropwhile`-style method you mentioned to look for any event that is one of the ones returned from the validator?  How does that sound?